One-Step Key Derivation Method with KMAC

FAlbertDev opened the pull request:
As mentioned in PR #4119, this pull request introduces the One-Step Key Derivation Method (KDM) of [SP800-56Cr2](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Cr2.pdf) Section 4 using KMAC.

Previously, the One-Step KDM was defined in [SP800-56Ar2](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar2.pdf). However, with the revision [SP800-56Ar3](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar3.pdf), it has been moved to [SP800-56Cr2](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Cr2.pdf). To reflect this change, I have updated comments and internal interfaces accordingly. The public API and external interfaces, including module names, remain unchanged.

Additionally, this pull request includes a KMAC option for the one-step Key Derivation Function (KDF), which is utilized by the Ounsworth Key Encapsulation Mechanism (KEM) combiner. Also, I added the test vectors I got for the Ounsworth PR. These include tests for the KMAC option and some for SHA-3 instances, which were not yet covered.

randombit commented on src/tests/data/kdf/sp800_56a.vec:
```diff
@@ -2411,3 +2414,342 @@ Salt =
+
+# Group 58 - Test 286
+Salt = 0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
+Secret = 19CFD413A6B255735D202A3FCD5F7BF3CEF007836227161DB48D7FDF88FAE72D8D1E8B099CFBEE9306348D735CE408EA28C54117F772D2A4D8700E94455BA8EA948724D549A4B90F08FEF634CBEF8EC222C43B30E57BB4B1F4D5AE560AF691FC695962EAA72AAF64F62693AC24C67986C6188BD6581243F771CB00F908B0359BAA1BFB457555DBAD9B93422972267CF4CEB60640923C17659ED83F82D84BF7395005D2A1BB483DCD591D136F77E2CFB14DD84400E985A17DBCB68DB32234CBAE03BA0C ...
```
Some of these test vectors are really long and I don't see much justification for them considering nothing about the KDF varies on the input lengths

FAlbertDev replied:
Agreed. I removed all vectors with excessive lengths!

randombit commented on src/lib/kdf/sp800_56a/sp800_56c_one_step.cpp:
```diff
@@ -0,0 +1,192 @@
+                  std::span<const uint8_t> z,
+                  std::span<const uint8_t> fixed_info,
+                  Buffered_Computation& h,
+                  const std::function<void(Buffered_Computation*)>& reset_h_callback) {
```
Seems better to template this over a `T` to avoid a `dynamic_cast` for each iteration of the loop.

Also `Buffered_Computation` as a base class is anyway deprecated so we'd have to rewrite this when it's removed.

FAlbertDev replied:
I didn't know Buffered_Computation is deprecated. I'll use a template.

randombit commented on src/lib/kdf/sp800_56a/sp800_56c_one_step.cpp:
```diff
@@ -0,0 +1,192 @@
+      //     OtherInput, serves as the HMAC#/KMAC# key ..
+
+      // SP 800-56A specifies if the salt is empty then a block of zeros
+      // equal to the hash's underlying block size are used. However this
```
This is an issue in the original comment but we should fix it while we're here

> However *for HMAC* this is equivalent

Since that's not generally true for eg KMAC or whatever

randombit commented on src/lib/kdf/sp800_56a/sp800_56c_one_step.cpp:
```diff
@@ -0,0 +1,192 @@
+      // equal to the hash's underlying block size are used. However this
+      // is equivalent to setting a zero-length key, so the same call
+      // works for either case.
+      kdf_mac->set_key(std::span{salt, salt_len});
```
Duplicated call to `set_key`

randombit commented on src/lib/kdf/sp800_56a/sp800_56c_one_step.h:
```diff
@@ -98,12 +100,73 @@ class SP800_56A_HMAC final : public KDF {
+      * @param label label for the expansion step
+      * @param label_len size of label in bytes
+      *
+      * @throws Invalid_Argument key_len > 2^32 or MAC is not a HMAC
```
Incorrect comment re HMAC

FAlbertDev replied:
This does not even check if MAC is an HMAC. This check is performed in the constructor. The max key length is also wrong 😅

randombit commented on src/lib/kdf/sp800_56a/sp800_56c_one_step.h:
```diff
@@ -98,12 +100,73 @@ class SP800_56A_HMAC final : public KDF {
+
+      std::unique_ptr<KDF> new_object() const override { return std::make_unique<SP800_56C_One_Step_KMAC128>(); }
+
+   protected:
```
Why not `private`?

randombit commented on src/lib/kdf/sp800_56a/sp800_56c_one_step.h:
```diff
@@ -98,12 +100,73 @@ class SP800_56A_HMAC final : public KDF {
+
+      std::unique_ptr<KDF> new_object() const override { return std::make_unique<SP800_56C_One_Step_KMAC256>(); }
+
+   protected:
```
`private`?

randombit commented on src/lib/kdf/sp800_56c/sp800_56c_two_step.cpp:
```diff
@@ -0,0 +1,41 @@
+/*
```
Why is `git`s rename logic so stupid

randombit reviewed (commented):
Left some comments but overall seems fine

FAlbertDev commented:
Thanks, @randombit, for your input! It seems I left some copy-past issues in the new implementation. Also, I did not correctly handle default_salts for KMAC. The doxygen comments were also deprecated in various places.

I addressed your suggestions and the other mentioned issues in the latest commit.
